### PR TITLE
Add driver for ADS114S08

### PR DESCRIPTION
**Pull request by benediktibk**
Add a driver for the TI ADS114S08 16-bit ADC

### CODE REVIEW

**Review by x64x2 — DISMISSED** [review]

File: drivers/adc/adc_ads114s08.c [hunk 1]
```diff
@@ -0,0 +1,906 @@
...
+	ADS114_REGISTER_GPIODAT = 0x10,
+	ADS114_REGISTER_GPIOCON = 0x11,
+};
+
+#define ADS114_REGISTER_GENERATE_BIT_MASK(length) ((1 << length) - 1)
+#define ADS114_REGISTER_GET_VALUE(value, pos, length)                                              \
+	((value >> pos) & ADS114_REGISTER_GENERATE_BIT_MASK(length))
+#define ADS114_REGISTER_SET_VALUE(target, value, pos, length)                                      \
+	target &= ~(ADS114_REGISTER_GENERATE_BIT_MASK(length) << pos);                             \
+	target |= ((value & ADS114_REGISTER_GENERATE_BIT_MASK(length)) << pos)
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:78]: Please examine if the macros in `include/zephyr/sys/util.h` can be used here (`FIELD_PREP()`, `FIELD_GET()`, `GENMASK()`, ...)

File: drivers/adc/adc_ads114s08.c [hunk 2]
```diff
@@ -0,0 +1,906 @@
...
+
+	if (result != 0) {
+		return result;
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_PGA, gain);
+
+	if (result != 0) {
+		return result;
+	}
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:544]: Some debug logs would be nice here.

File: drivers/adc/adc_ads114s08.c [hunk 3]
```diff
@@ -0,0 +1,906 @@
...
+
+	result = gpio_pin_set_dt(&config->gpio_start_sync, 1);
+
+	if (result != 0) {
+		LOG_ERR("unable to start ADC operation");
+		return result;
+	}
+
+	k_sleep(K_USEC(ADS114_START_SYNC_PULSE_DURATION_IN_US + ADS114_SETUP_TIME_IN_US));
+	gpio_pin_set_dt(&config->gpio_start_sync, 0);
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:637]: No support for START command via SPI?

File: drivers/adc/adc_ads114s08.c [hunk 4]
```diff
@@ -0,0 +1,906 @@
...
+	int result;
+	const struct ads114_config *config = dev->config;
+	struct ads114_data *data = dev->data;
+
+	adc_context_init(&data->ctx);
+
+	k_sem_init(&data->data_ready_signal, 0, 1);
+	k_sem_init(&data->acquire_signal, 0, 1);
+
+	if (!device_is_ready(config->bus.bus)) {
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:789]: Please use `spi_is_ready_dt()`.

File: drivers/adc/adc_ads114s08.c [hunk 5]
```diff
@@ -0,0 +1,906 @@
...
+	k_thread_name_set(tid, "adc_ads114");
+#endif
+
+	k_busy_wait(ADS114_RESET_LOW_TIME_IN_US);
+	gpio_pin_set_dt(&config->gpio_reset, 0);
+	k_busy_wait(ADS114_RESET_DELAY_TIME_IN_US);
+
+	result = ads114_read_register(&config->bus, ADS114_REGISTER_STATUS, &status);
+	if (result != 0) {
+		return result;
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:840]: An error log would be nice.

File: drivers/adc/adc_ads114s08.c [hunk 6]
```diff
@@ -0,0 +1,906 @@
...
+
+	/*
+	 * Activate internal voltage reference during initialization to
+	 * avoid the necessary setup time for it to settle later on.
+	 */
+	ADS114_REGISTER_REF_SET_DEFAULTS(reference_control);
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_REF, reference_control);
+	if (result != 0) {
+		return result;
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:856]: Same here.

File: drivers/adc/adc_ads114s08.c [hunk 7]
```diff
@@ -0,0 +1,906 @@
...
+	}
+
+	/*
+	 * Ensure that the internal voltage reference is active.
+	 */
+	result = ads114_read_register(&config->bus, ADS114_REGISTER_REF, &reference_control_read);
+	if (result != 0) {
+		return result;
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_ads114s08.c:864]: And here.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/adc/CMakeLists.txt [hunk 8]
```diff
@@ -28,6 +28,7 @@ zephyr_library_sources_ifdef(CONFIG_ADC_TEST		adc_test.c)
 zephyr_library_sources_ifdef(CONFIG_ADC_ADS1X1X		adc_ads1x1x.c)
 zephyr_library_sources_ifdef(CONFIG_ADC_GD32		adc_gd32.c)
 zephyr_library_sources_ifdef(CONFIG_ADC_ADS1119		adc_ads1119.c)
+zephyr_library_sources_ifdef(CONFIG_ADC_ADS114S08		adc_ads114s08.c)
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/CMakeLists.txt:31]: I actually thought about it but decided against, as I do not have the hardware available. Hence, I wouldn't be able to test it.

File: dts/bindings/adc/ti,ads114s08.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    specifier-space: gpio
+
+  gpio-start-sync:
+    type: phandle-array
+    required: true
+    description: "GPIO for start/sync"
+    specifier-space: gpio
+
+io-channel-cells:
+  - input
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply dts/bindings/adc/ti,ads114s08.yaml:33]: Yes, there is differential input support. Looks like I started with a copy of an ADC which only supported single ended input.

Anyway, changing this to positive and negative also implies changing the value for io-channel-cells to 2, if I understood this correctly? But how is this than reference in the device tree? So far, for my test project of this driver, I have used the following:

```
 / {
	zephyr,user {
		/* adjust channel number according to pinmux in board.dts */
		io-channels = <&spi_ads114s08 0>, <&spi_ads114s08 1>, <&spi_ads114s08 2>, <&spi_ads114s08 3>, <&spi_ads114s08 4>;
	};
};

&spi1 {
	status = "okay";
	/* one entry for every devices at spi.dtsi */
	cs-gpios = <&gpiod 14 GPIO_ACTIVE_LOW>;

	spi_ads114s08: ads114s08@0 {
		status = "okay";
		compatible = "ti,ads114s08";
		spi-max-frequency = <10000000>;
		reg = <0x0>;	
		#address-cells = <1>;
		#size-cells = <0>;
		#io-channel-cells = <2>;
		reset-gpios = <&gpioa 15 GPIO_ACTIVE_LOW>;
		drdy-gpios = <&gpioc 7 GPIO_ACTIVE_LOW>;
		start-sync-gpios = <&gpiob 5 GPIO_ACTIVE_LOW>;
	
		channel@0 {
			reg = <0>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <16>;
			zephyr,input-positive = <0>;
		};
	
		channel@1 {
			reg = <1>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <16>;
			zephyr,input-positive = <1>;
		};
	
		channel@2 {
			reg = <2>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_INTERNAL";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <16>;
			zephyr,input-positive = <0>;
			zephyr,input-negative = <1>;
		};
	
		channel@3 {
			reg = <3>;
			zephyr,gain = "ADC_GAIN_1";
			zephyr,reference = "ADC_REF_EXTERNAL0";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <16>;
			zephyr,input-positive = <0>;
			zephyr,input-negative = <1>;
			zephyr,vref-mv = <1571>;
		};
	
		channel@4 {
			reg = <4>;
			zephyr,gain = "ADC_GAIN_2";
			zephyr,reference = "ADC_REF_EXTERNAL0";
			zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
			zephyr,resolution = <16>;
			zephyr,input-positive = <0>;
			zephyr,input-negative = <1>;
			zephyr,vref-mv = <1571>;
		};
	};
};
```

How would the channels than be specified with this? Unfortunately, I couldn't find an example for this. On top, what is then the purpose of input-positive and input-negative?

File: dts/bindings/adc/ti,ads114s08.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+compatible: "ti,ads114s08"
+
+include: [adc-controller.yaml, spi-device.yaml]
+
+properties:
+  "#io-channel-cells":
+    const: 1
+
+  gpio-reset:
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply dts/bindings/adc/ti,ads114s08.yaml]: done

File: dts/bindings/adc/ti,ads114s08.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    type: phandle-array
+    required: true
+    description: "GPIO for reset"
+    specifier-space: gpio
+
+  gpio-data-ready:
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply dts/bindings/adc/ti,ads114s08.yaml]: done

File: dts/bindings/adc/ti,ads114s08.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    type: phandle-array
+    required: true
+    description: "GPIO for data ready"
+    specifier-space: gpio
+
+  gpio-start-sync:
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply dts/bindings/adc/ti,ads114s08.yaml]: done

File: drivers/adc/adc_ads114s08.c (hunk 1 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c]: done

File: drivers/adc/adc_ads114s08.c (hunk 2 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c]: done

File: drivers/adc/adc_ads114s08.c (hunk 3 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c]: Yes, there is one. I haven't implemented it as I have the signal available in my design. Added the possibility to use the driver without the dedicated START/SYNC signal now anyway.

File: drivers/adc/adc_ads114s08.c (hunk 4 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c]: done

File: drivers/adc/adc_ads114s08.c (hunk 5 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c:893]: done

File: drivers/adc/adc_ads114s08.c (hunk 6 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c:910]: done

File: drivers/adc/adc_ads114s08.c (hunk 7 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c:919]: done

**Review by cfriedt — DISMISSED** [review]
LGTM

**Comment by DarrylGamroth** [comment]
TI uses the naming ADS114S0x and ADS124S0x for 16 and 24 bit versions of this ADC, where the x is the number of channels:
ADS114S08 - 16 bit, 8 channel
ADS114S06 - 16 bit, 6 channel
ADS124S08 - 24 bit, 8 channel
ADS124S06 - 24 bit, 6 channel

The Linux driver uses the prefix ads124s_ instead of ads114_. I'd recommend adding the 's' (ads114s_) as as the ADS114x is a different part family that is not compatible.

File: drivers/adc/adc_ads114s08.c [hunk 13]
```diff
@@ -0,0 +1,969 @@
...
+		LOG_ERR("unable to configure data rate");
+		return result;
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_PGA, gain);
+
+	if (result != 0) {
+		LOG_ERR("unable to configure PGA");
+		return result;
+	}
```
Comment (DarrylGamroth) [inline drivers/adc/adc_ads114s08.c:578]: These writes can be combined into a single multi-register write by writing these registers in order (INPMUX, PGA, DATARATE, REF) by extending ads114_write_register() to support a length parameter.

Refer to [lmp90xxx_adc_channel_setup()](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_lmp90xxx.c#L458) for an example.

Comment (benediktibk) [reply drivers/adc/adc_ads114s08.c]: Good idea, I just modified the approach slightly to ensure that the caller has an assert check that the correct registers are written to.

**Comment by benediktibk** [comment]
> TI uses the naming ADS114S0x and ADS124S0x for 16 and 24 bit versions of this ADC, where the x is the number of channels: ADS114S08 - 16 bit, 8 channel ADS114S06 - 16 bit, 6 channel ADS124S08 - 24 bit, 8 channel ADS124S06 - 24 bit, 6 channel
> 
> The Linux driver uses the prefix ads124s_ instead of ads114_. I'd recommend adding the 's' (ads114s_) as the ADS114x is a different part family that is not compatible.

@DarrylGamroth I took a look at the datasheet of the non-s version as well, this really is kinda confusing. But I think @henrikbrixandersen and I agreed upon only implementing the ADS114S08 at the moment, as this is the only hardware I have available and can test the driver on. Hence, I think there is nothing around without the s? Or have I missed something?

**Comment by DarrylGamroth** [comment]
> > TI uses the naming ADS114S0x and ADS124S0x for 16 and 24 bit versions of this ADC, where the x is the number of channels: ADS114S08 - 16 bit, 8 channel ADS114S06 - 16 bit, 6 channel ADS124S08 - 24 bit, 8 channel ADS124S06 - 24 bit, 6 channel
> > The Linux driver uses the prefix ads124s_ instead of ads114_. I'd recommend adding the 's' (ads114s_) as the ADS114x is a different part family that is not compatible.
> 
> @DarrylGamroth I took a look at the datasheet of the non-s version as well, this really is kinda confusing. But I think @henrikbrixandersen and I agreed upon only implementing the ADS114S08 at the moment, as this is the only hardware I have available and can test the driver on. Hence, I think there is nothing around without the s? Or have I missed something?

@benediktibk Sorry, what I meant is to reduce confusion by using the namespace 'ads114s_' or 'ads114s0x_' rather than 'ads114_' for this driver.

If someone wants to write a driver for the unrelated ADS114X family in the future they could use the namespace 'ads114x_'

**Comment by benediktibk** [comment]
> > > TI uses the naming ADS114S0x and ADS124S0x for 16 and 24 bit versions of this ADC, where the x is the number of channels: ADS114S08 - 16 bit, 8 channel ADS114S06 - 16 bit, 6 channel ADS124S08 - 24 bit, 8 channel ADS124S06 - 24 bit, 6 channel
> > > The Linux driver uses the prefix ads124s_ instead of ads114_. I'd recommend adding the 's' (ads114s_) as the ADS114x is a different part family that is not compatible.
> > 
> > 
> > @DarrylGamroth I took a look at the datasheet of the non-s version as well, this really is kinda confusing. But I think @henrikbrixandersen and I agreed upon only implementing the ADS114S08 at the moment, as this is the only hardware I have available and can test the driver on. Hence, I think there is nothing around without the s? Or have I missed something?
> 
> @benediktibk Sorry, what I meant is to reduce confusion by using the namespace 'ads114s_' or 'ads114s0x_' rather than 'ads114_' for this driver.
> 
> If someone wants to write a driver for the unrelated ADS114X family in the future they could use the namespace 'ads114x_'

I think I got it now, renamed all the functions, defines, ... etc.

**Review by cfriedt — APPROVED** [review]

**Comment by benediktibk** [comment]
@henrikbrixandersen Would you please mind taking another look?

**Review by cfriedt — DISMISSED** [review]

**Comment by benediktibk** [comment]
Resolved merge conflicts

**Comment by MaureenHelm** [comment]
@henrikbrixandersen @anangl @cfriedt please revisit

**Review by cfriedt — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

File: drivers/adc/adc_ads114s0x.c [hunk 14]
```diff
@@ -0,0 +1,1025 @@
...
+#define ADS114S0X_INPUT_SELECTION_AINCOM		    12
+#define ADS114S0X_RESOLUTION				    16
+#define ADS114S0X_REF_INTERNAL				    2500
+
+/*
+ * According to the datasheet this should be 4096 * 1/fclk = 1us,
+ * but it actually takes more time for the ADC to report ready in
+ * its status register.
+ */
+#define ADS114S0X_RESET_DELAY_TIME_IN_US 1000
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:35]: If you use the correct fclk value, which is 4.096 MHz (= 4096 kHz), then you'll get exactly 1 ms here.

File: drivers/adc/adc_ads114s0x.c [hunk 15]
```diff
@@ -0,0 +1,1025 @@
...
+
+#define ADS114S0X_RESET_LOW_TIME_IN_US                                                             \
+	(ADS114S0X_RESET_LOW_TIME_IN_CLOCK_CYCLES / ADS114S0X_CLK_FREQ_IN_MHZ)
+#define ADS114S0X_START_SYNC_PULSE_DURATION_IN_US                                                  \
+	(ADS114S0X_START_SYNC_PULSE_DURATION_IN_CLOCK_CYCLES / ADS114S0X_CLK_FREQ_IN_MHZ)
+#define ADS114S0X_SETUP_TIME_IN_US                                                                 \
+	(ADS114S0X_SETUP_TIME_IN_CLOCK_CYCLES / ADS114S0X_CLK_FREQ_IN_MHZ)
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:42]: These time values will currently get the value of 0 because of the incorrect frequency used.

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_ads114s0x.c [hunk 16]
```diff
@@ -0,0 +1,1025 @@
...
+	if (channel_cfg->differential) {
+		if (channel_cfg->input_positive >= ADS114S0X_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("positive channel input %i is invalid",
+				channel_cfg->input_positive);
+			return -EINVAL;
+		}
+
+		if (channel_cfg->input_negative >= ADS114S0X_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("negative channel input %i is invalid",
+				channel_cfg->input_positive);
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:555]:
```suggestion
		if (channel_cfg->input_negative >= ADS114S0X_INPUT_SELECTION_AINCOM) {
			LOG_ERR("negative channel input %i is invalid",
				channel_cfg->input_negative);
```

File: drivers/adc/adc_ads114s0x.c [hunk 17]
```diff
@@ -0,0 +1,1025 @@
...
+		if (channel_cfg->input_positive == channel_cfg->input_negative) {
+			LOG_ERR("negative and positive channel inputs must be different");
+			return -EINVAL;
+		}
+
+		ADS114S0X_REGISTER_INPMUX_MUXP_SET(input_mux, channel_cfg->input_positive);
+		ADS114S0X_REGISTER_INPMUX_MUXN_SET(input_mux, channel_cfg->input_negative);
+	} else {
+		if (channel_cfg->channel_id >= ADS114S0X_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("channel input %i is invalid", channel_cfg->input_positive);
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:568]: See:
https://github.com/zephyrproject-rtos/zephyr/blob/4a218aa1990dc7cdd4785d277a52f546cb4013e4/drivers/adc/adc_ads114s0x.c#L511-L513
```suggestion
		if (channel_cfg->input_positive >= ADS114S0X_INPUT_SELECTION_AINCOM) {
			LOG_ERR("channel input %i is invalid", channel_cfg->input_positive);
```

File: drivers/adc/adc_ads114s0x.c [hunk 18]
```diff
@@ -0,0 +1,1025 @@
...
+		(static K_KERNEL_STACK_DEFINE(                                                     \
+			 thread_stack_##n, CONFIG_ADC_ADS114S0X_ACQUISITION_THREAD_STACK_SIZE);))  \
+	static const struct ads114s0x_config config_##n = {                                        \
+		.bus = SPI_DT_SPEC_INST_GET(                                                       \
+			n, SPI_OP_MODE_MASTER | SPI_MODE_CPHA | SPI_WORD_SET(8), 0),               \
+		IF_ENABLED(CONFIG_ADC_ASYNC, (.stack = thread_stack_##n,))                         \
+		.gpio_reset = GPIO_DT_SPEC_GET_BY_IDX_OR(DT_DRV_INST(n), reset_gpios, 0, {0}),     \
+		.gpio_data_ready = GPIO_DT_SPEC_GET_BY_IDX(DT_DRV_INST(n), drdy_gpios, 0),         \
+		.gpio_start_sync = GPIO_DT_SPEC_GET_BY_IDX_OR(                                     \
+			DT_DRV_INST(n), start_sync_gpios, 0, {0}),                                 \
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:1019]:
```suggestion
		.gpio_reset = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {0}),			   \
		.gpio_data_ready = GPIO_DT_SPEC_INST_GET(n, drdy_gpios),			   \
		.gpio_start_sync = GPIO_DT_SPEC_INST_GET_OR(n, start_sync_gpios, {0}),		   \
```

File: drivers/adc/adc_ads114s0x.c [hunk 19]
```diff
@@ -0,0 +1,1025 @@
...
+		LOG_ERR("failed to add data ready callback");
+		return -EIO;
+	}
+
+#if CONFIG_ADC_ASYNC
+	const k_tid_t tid = k_thread_create(
+		&data->thread, config->stack, CONFIG_ADC_ADS114S0X_ACQUISITION_THREAD_STACK_SIZE,
+		(k_thread_entry_t)ads114s0x_acquisition_thread, (void *)dev, NULL, NULL,
+		CONFIG_ADC_ADS114S0X_ASYNC_THREAD_INIT_PRIO, 0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_ads114");
```
Comment (anangl) [inline drivers/adc/adc_ads114s0x.c:934]:
```suggestion
	k_thread_name_set(tid, "adc_ads114s0x");
```

File: drivers/adc/adc_ads114s0x.c [hunk 20]
```diff
@@ -0,0 +1,1025 @@
...
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER 1
+#include "adc_context.h"
+
+LOG_MODULE_REGISTER(ads114s0x, CONFIG_ADC_LOG_LEVEL);
+
+#define ADS114S0X_CLK_FREQ_IN_MHZ			    4096
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: Thanks for pointing this out, fixed it.

File: drivers/adc/adc_ads114s0x.c (hunk 14 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

File: drivers/adc/adc_ads114s0x.c (hunk 15 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

File: drivers/adc/adc_ads114s0x.c (hunk 16 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

File: drivers/adc/adc_ads114s0x.c (hunk 17 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

File: drivers/adc/adc_ads114s0x.c (hunk 18 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

File: drivers/adc/adc_ads114s0x.c (hunk 19 above)
Comment (benediktibk) [reply drivers/adc/adc_ads114s0x.c]: done

**Comment by benediktibk** [comment]
On top of the requested changes (btw, thanks again for the input) I stumbled upon an issue during a power on with the reset time. I've added a safety margin which seems to be necessary after the power supply has stabilized. But I'm uncertain if this should better be a Kconfig option?

**Review by cfriedt — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Comment by purpl3F0x** [comment]
Should support for ads124s0x be added in this driver or a sepperate one ? The chips are practically identical, only thing that differs are the 24bit data and the Calibration value registers being 3x8bit.

**Comment by benediktibk** [comment]
@purpl3F0x This pull request is already merged. But feel free to extend the driver in another PR.

**Pull request merged**